Note if mismatched types have a similar name

mejrs opened the pull request:
If users get a type error between similarly named types, it will point out that these are actually different types, and where they were defined.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @fee1-dead (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

jyn514 commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
-            match (&terr, expected == found) {
-                (TypeError::Sorts(values), extra) => {
+            let looks_similar = |e: ExpectedFound<Ty<'_>>| {
+                // We're only interested in adts
+                if let (Some(e), Some(f)) = (e.expected.ty_adt_def(), e.found.ty_adt_def()) {
```
Why limit this to ADTs? This can also be true for ADTs named after primitives: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=166bd629e2141b0cd76b038fc470c39a

mejrs replied:
Excellent point, thanks.

jyn514 commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                    let found_name = values.found.sort_string(self.tcx);
+                    let expected_name = values.expected.sort_string(self.tcx);
+
+                    diag.note(format!("{found_name} and {expected_name} have similar names, but are actually distinct types"));
```
how does this interact with the existing "help: maybe two different crates are being used?" message? IMO we should probably only show one or the other; it would be nice to add a test case.

mejrs replied:
It shows both; see `src/test/ui/type/type-mismatch-same-crate-name`. Its output got updated with this PR.

jyn514 commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                        let def_span = self.tcx.def_span(defid);
+
+                        let msg = if defid.is_local() {
+                            format!("{name} is defined in the current crate.")
```
using `.` in errors in uncommon; does the message still make sense without?

jyn514 replied:
this doesn't seem helpful when both ADTs are defined in the current crate; can we highlight the module name maybe?

jyn514 commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1670,8 +1671,51 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                        {
+                            // if it's a diagnostic item, it's definitely defined in std/core/alloc
+                            // otherwise might be, might not be.
+                            format!("{name} is defined in the standard library.")
+                        } else {
+                            let crate_name = self.tcx.crate_name(defid.krate);
+                            format!("{name} is defined in crate `{crate_name}`.")
```
having this inconsistency between libstd types that are diagnostic items and types that aren't seems unfortunate. I think we should either switch on `crate_name() == std/core/alloc` or use `crate {name}` consistently. My preference is for the latter; if two types with the same name are defined in core and std, this message will not be particularly helpful.

mejrs replied:
Let's do the latter :)

fee1-dead commented:
@rustbot author

mejrs commented:
@rustbot ready

jyn514 commented on src/test/ui/fully-qualified-type/fully-qualified-type-name2.stderr:
```diff
@@ -7,12 +7,12 @@ LL |     return x;
    |
    = note: enum `x::Foo` and enum `y::Foo` have similar names, but are actually distinct types
-note: enum `x::Foo` is defined in the current crate.
+note: enum `x::Foo` is defined in module ::x of the current crate
```
In normal rust code, this means "the external crate x"; having it be different from what the diagnostic means seems confusing.
```suggestion
note: enum `x::Foo` is defined in module `crate::x` of the current crate
```

mejrs replied:
good point. I can think of some ways to word it:
- ```note: enum `x::Foo` is defined in module `crate::x` ```
- ```note: enum `x::Foo` is defined in module `whatever_the_crate_name::x` ```
 - ```note: enum `x::Foo` is defined in module `crate::x` of the current crate```
- ```note: enum `x::Foo` is defined in module `whatever_the_crate_name::x` of the current crate```

Personally I'm leaning towards the second, but I am unsure what is clearest to a new user.

jyn514 replied:
The second has the same problem though - crates don't know their own names, so `whatever_the_crate_name::x` won't compile if you copy it into the source code.

mejrs replied:
Ohh, that's true. I forgot people might want to copy paste it. I have updated it to print as you suggested above.

bors commented:
:umbrella: The latest upstream changes (presumably #101938) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3075355077/jobs/4968704169) [(plain)](https://github.com/rust-lang/rust/commit/55d39b136c0296751a8329d2b97e944eee38a886/checks/8411420351/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

mejrs commented:
@rustbot ready

fee1-dead commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1653,8 +1654,157 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                PrimitiveExpected(ty::AdtDef<'tcx>, Ty<'tcx>),
+            }
+
+            let primitive_sym = |kind: &_| match kind {
+                ty::Bool => Some(sym::bool),
+                ty::Char => Some(sym::char),
+                ty::Float(f) => match f {
+                    ty::FloatTy::F32 => Some(sym::f32),
+                    ty::FloatTy::F64 => Some(sym::f64),
+                },
+                ty::Int(f) => match f {
+                    ty::IntTy::Isize => Some(sym::isize),
+                    ty::IntTy::I8 => Some(sym::i8),
+                    ty::IntTy::I16 => Some(sym::i16),
+                    ty::IntTy::I32 => Some(sym::i32),
+                    ty::IntTy::I64 => Some(sym::i64),
+                    ty::IntTy::I128 => Some(sym::i128),
+                },
+                ty::Uint(f) => match f {
+                    ty::UintTy::Usize => Some(sym::usize),
+                    ty::UintTy::U8 => Some(sym::u8),
+                    ty::UintTy::U16 => Some(sym::u16),
+                    ty::UintTy::U32 => Some(sym::u32),
+                    ty::UintTy::U64 => Some(sym::u64),
+                    ty::UintTy::U128 => Some(sym::u128),
+                },
+                _ => None,
+            };
```
This should be an associated function for `TyKind`. Should be okay for now, because it can be in a followup.

mejrs replied:
I don't think it can be an associated function on TyKind (or Symbol for that matter) because that leads to cyclic dependencies. But I'm happy to make it a free function/trait somewhere, or an associated function on `rustc_middle::ty::Ty`

fee1-dead commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1653,8 +1654,157 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                match (fk, ek) {
+                    (
+                        ty::Adt(adt, _),
+                        ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_),
```
This could use `is_primitive` instead.

fee1-dead commented on compiler/rustc_infer/src/infer/error_reporting/mod.rs:
```diff
@@ -1653,8 +1654,157 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                        None
+                    }
+                    (
+                        ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_),
```
same above. I think you can also use an `if-let` guard to extract the primitive sym in this match statement directly.

fee1-dead commented on src/test/ui/mismatched_types/similar_paths.rs:
```diff
@@ -0,0 +1,11 @@
+enum Option<T>{
+    Some(T),
+    None,
+}
+
+pub fn foo() -> Option<u8>{
```
```suggestion
enum Option<T> {
    Some(T),
    None,
}

pub fn foo() -> Option<u8> {
```

fee1-dead commented on src/test/ui/mismatched_types/show_module.rs:
```diff
@@ -0,0 +1,18 @@
+pub mod blah{
+    pub mod baz{
+        pub struct Foo;
+    }
+}
+
+pub mod meh{
+    pub struct Foo;
+}
+
+pub type Foo = blah::baz::Foo;
+
+fn foo() -> Foo {
+    meh::Foo
+    //~^ ERROR mismatched types [E0308]
+}
+
+fn main(){}
```
```suggestion
pub mod blah {
    pub mod baz {
        pub struct Foo;
    }
}

pub mod meh {
    pub struct Foo;
}

pub type Foo = blah::baz::Foo;

fn foo() -> Foo {
    meh::Foo
    //~^ ERROR mismatched types [E0308]
}

fn main() {}
```

fee1-dead reviewed (commented):
Left a few nits. Should be good to go after reviews are addressed. Sorry for the late review!

mejrs commented:
Thank you for the review. I have:
- formatted the tests
- moved `primitive_sym` to a method on `rustc_middle::ty::Ty`
- (not your feedback but) changed the `Similar` enum to use named fields rather than tuple fields
- changed the rest of the code to work using the above.

fee1-dead reviewed (approved):
LGTM! Thanks

fee1-dead commented:
@bors r+

bors commented:
:pushpin: Commit c65866000e07270817465f3cbb8b1d01ecdb9a20 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead c65866000e07270817465f3cbb8b1d01ecdb9a20 -->
<!-- homu: {"type":"Approved","sha":"c65866000e07270817465f3cbb8b1d01ecdb9a20","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
